### DISCUSSION TITLE
Create or Modifiy Notes Depending on a Note Type

### DISCUSSION THREAD

**Discussion by MonikaLobinger** [SilentVoid13/Templater#1071] [Templates Showcase]
### Update 2023-04-06
[current version](https://github.com/SilentVoid13/Templater/discussions/1071#discussioncomment-5544305)
### What script does
It creates/modifies files according to following convention:
- Having some note types like "book" or "person" or "quote"
- There is a naming convention, e.g."{b}author - title", "lastname, firstname", "°some name author".
- The frontmatter attributes "aliases", "date_created", "tags", "publish", "cssclass" should be created or changed, depending on type and file name.
- If applied to an existing file other existing frontmatter attributes should stay.
### Was das Script tut
Es erzeugt/verändert Notizen entsprechend der folgenden Richtlinien:
- Es gibt Notiz Typen, wie "buch", "person" oder "zitat"
- Es gibt eine Namenskonvention, wie "{b}Autor - Titel", "Nachname, Vorname", "°irgendein name Autor"
- Die frontmatter Attribute "aliases", "date_created", "tags", "publish", "cssclass"  sollen entsprechend dem Dateitypen und dem Dateinamen erstellt oder angepasst werden.
- Wenn auf eine existierende Notiz angewandt, sollen andere Frontmatter Attribute bestehen bleiben.

### German and no JS Programmer
- The script is commented in German
- The script is not sophisticated JS

### Embedded
```
---
<%*
/* NUTZER Variablen: */
/* NUTZER: erstellte oder ergänzte frontmatter Werte  */
let aliases      = tp.frontmatter.aliases;
let date_created = tp.frontmatter.date_created;
let tags         = tp.frontmatter.tags;
let publish      = tp.frontmatter.publish;
let cssclass     = "";
/* NUTZER: weitere */
let foto         = "";
let anderepaare  = "";

/** Erstellt die obigen frontmatter Einträge mit Werten abhängig von 
 * einem Dateityp.
 * Mögliche Dateitpyen sind nach dem Kommentar in 'typen' festgelegt.
 *
 * Dieses Template kann bei der Erzeugung neuer Notizen ausgeführt 
 * werden.
 * Dazu kann `Templater: Create new Note from Template` (`Alt-N`) 
 * ausgeführt werden oder das Template wird als folder template 
 * verwendet.
 * Dieses Template kann innerhalb bestehender Notizen ausgeführt 
 * werden. Dazu wird `Templater: Open insert template modal`(`Alt-E`) 
 * ausgeführt.
 * Der alte frontmatter header muß vorhanden sein. Der Cursor muss 
 * an der Stelle stehen, wo das neue Template eingefügt werden soll. 
 * Danach kann der alte frontmatter header gelöscht werden.
 * 
 * Bei neuen Dateien werden die frontmatter Einträge neu erstellt. 
 * Dazu wird der Dateityp und der Dateiname abgefragt. 
 * Der Dateiname wird als Name der neuen Datei gesetzt.
 * 
 * Bei vorhandenen Dateien werden die oben angegebenen frontmatter 
 * erstellt, ergänzt, verändert wie im nächsten Absatz angegeben.
 * Alle anderen frontmatter Einträge werden übernommen.
 * 
 * 
 * - Der erstellte/verwendete Eintrag in `aliases` leitet sich vom 
 *   Dateinamen ab:
 *   der kenner vorne wird entfernt. Alle Kommas werden entfernt. 
 *   Beim Dateityp "person" wird der Teil vor dem ersten Komma 
 *     der letzte Teil. 
 *     So wird "Meier, Antonia" zu "Antonia Meier"
 *   beim Dateityp "ort" wird der Teil nach dem ersten Komma 
 *     in runde Klammern eingeschlossen.
 *     So wird "Ort, Deutschland" zu "Ort(Deutschland)"
 * - 'date_created' wird beibehalten, falls vorhanden, sonst wird es 
 *   das aktuelle Datum
 * - Das erstellte/beibehaltene `tag` in `tags` ist ein geschachteltes 
 *   Tag aus 'haupttag' und Dateityp.
 * - 'publish' wird beibehalten, bei Neuerstellung ist der Wert 
 *   `false`
 * - `cssclass` wird immer der Dateityp, eventuell existierende Werte
 *   werden überschrieben
 */

/* Voreinstellungen */
let haupttag     = "0";
let typen        = ["audio", "buch", "ort", "person", "video", "web", "zitat", "zitate"];
let kein_kenner  = ["ort", "person"];

let audioprompt  = "?Podcast/Reihe - Autornachname - Audiotitel";
let buchprompt   = "Autornachname - Buchtitel";
let ortprompt    = "Ortsname, Land";
let personprompt = "Personnachname, Personvorname ?Geburtsdatum";
let videoprompt  = "?Reihe - ?Autornachname - Videotitel";
let webprompt    = "?Autor - Webseitentitel - ?Datum";
let zitatprompt  = "Titel Autornachname";
let zitateprompt = "Titel Autornachname";

let audio_kenner = "{a}";
let buch_kenner  = "{b}";
let ort_kenner   = "";
let person_kenner= "";
let video_kenner = "{v}";
let web_kenner   = "{w}";
let zitat_kenner = "°";
let zitate_kenner= "°°";

let audio_foto   = "pexels-foteros-352505_200.jpg";
let buch_foto    = "pexels-gül-işık-2203051_200.jpg";
let ort_foto     = "pexels-dzenina-lukac-1563005_200.jpg";
let person_foto  = "pexels-lucas-andrade-14097235_200.jpg";
let video_foto   = "pexels-vlad-vasnetsov-2363675_200.jpg";
let web_foto     = "pexels-sururi-ballıdağ-_200.jpeg";
let zitat_foto   = "";
let zitate_foto  = "";

/* Interne Variablen */
let dateityp     = "unbekannt"
let fname        = tp.file.title;
let kenner_len   = 0;
let aliasname    = "";

/* ** Berechnen von `dateityp` und `fname` ** */
/* Neue Dateien */
if(fname.startsWith("Unbenannt") || fname.startsWith("Untitled")) {
  let title  = "";
  dateityp = await tp.system.suggester(typen, typen);
  if(dateityp == "audio") {
    fname = await tp.system.prompt(audioprompt);
    title = audio_kenner + fname;
  } else if (dateityp == "buch") {
    fname = await tp.system.prompt(buchprompt);
    title = buch_kenner + fname;
  } else if (dateityp == "ort") {
    fname = await tp.system.prompt(ortprompt);
    title = ort_kenner + fname;
  } else if (dateityp == "person") {
    fname = await tp.system.prompt(personprompt);
    title = person_kenner + fname;
  } else if (dateityp == "video") {
    fname = await tp.system.prompt(videoprompt);
    title = video_kenner + fname;
  } else if (dateityp == "web") {
    fname = await tp.system.prompt(webprompt);
    title = web_kenner + fname;
  } else if (dateityp == "zitat") {
    fname = await tp.system.prompt(zitatprompt);
    title = zitat_kenner + fname;
  } else if (dateityp == "zitate") {
    fname = await tp.system.prompt(zitateprompt);
    title = zitate_kenner + fname;
  }
  await tp.file.rename(title);
} 
/* existierende Dateien */
else {
  if(audio_kenner.length > 0 && fname.startsWith(audio_kenner)) {
    dateityp = "audio"; 
    kenner_len = audio_kenner.length;
  } else if(buch_kenner.length > 0 && fname.startsWith(buch_kenner)) {
    dateityp = "buch"; 
    kenner_len = buch_kenner.length;
  } else if(ort_kenner.length > 0 && fname.startsWith(ort_kenner)) {
    dateityp = "ort"; 
    kenner_len = ort_kenner.length;
  } else if(person_kenner.length > 0 && fname.startsWith(person_kenner)) {
    dateityp = "person"; 
    kenner_len = person_kenner.length;
  } else if(video_kenner.length > 0 && fname.startsWith(video_kenner)) {
    dateityp = "video"; 
    kenner_len = video_kenner.length;
  } else if(web_kenner.length > 0 && fname.startsWith(web_kenner)) {
    dateityp = "web"; 
    kenner_len = web_kenner.length;
  } else if(zitat_kenner.length > 0 && fname.startsWith(zitat_kenner)) {
    dateityp = "zitat"; 
    kenner_len = zitat_kenner.length;
  } else if(zitate_kenner.length > 0 && fname.startsWith(zitate_kenner)) {
    dateityp = "zitate"; 
    kenner_len = zitate_kenner.length;
  } else {
    dateityp = await tp.system.suggester(kein_kenner, kein_kenner);
  }
  fname = fname.substring(kenner_len); 
}

/* ** Berechnen von `aliases` ** */
/* Kommas entfernen */
if(dateityp != "ort" && dateityp != "person") {
  aliasname = await fname.replace(/,/g,` `).replace(/  /g,` `);
} else { // ort, person
  aliasname = await fname.replace(/, /g,`,`);
  let strArr = aliasname.split(",");
  aliasname = strArr[0];
  strArr.shift();
  if(dateityp == "ort") {
    aliasname += "(" + strArr.join(' ') + ")";
  } else if(dateityp == "person") {
    aliasname = strArr.join(' ') + " " + aliasname;
  }
}
if(typeof aliases === 'undefined') {
  aliases = aliasname;
} else if(!aliases.contains(aliasname)) {
  aliases = aliasname + ", " + aliases;
}

/* ** Berechnen von `date_created` ** */
if(typeof date_created === 'undefined') {
  date_created = tp.date.now();
}

/* ** Berechnen von `tags` ** */
if(typeof tags === 'undefined') {
  tags = haupttag + "/"+dateityp;
} else if(!tags.contains("0/"+dateityp)) {
  tags = haupttag + "/"+dateityp + " " + tags;
}

/* ** Berechnen von `publish` ** */
if(typeof publish === 'undefined') {
  publish = false;
}

/* ** Berechnen von `cssclass` ** */
cssclass = dateityp;

/* ** Berechnen von `foto` ** */
if(dateityp == "audio") {
  if(audio_foto.length > 0) {
    foto = audio_foto;
  }
} else if(dateityp == "buch") {
  if(buch_foto.length > 0) {
    foto = buch_foto;
  }
} else if(dateityp == "ort") {
  if(ort_foto.length > 0) {
    foto = ort_foto;
  }
} else if(dateityp == "person") {
  if(person_foto.length > 0) {
    foto = person_foto;
  }
} else if(dateityp == "video") {
  if(video_foto.length > 0) {
    foto = video_foto;
  }
} else if(dateityp == "web") {
  if(web_foto.length > 0) {
    foto = web_foto;
  }
} else if(dateityp == "zitat") {
  if(zitat_foto.length > 0) {
    foto = zitat_foto;
  }
} else if(dateityp == "zitate") {
  if(zitate_foto.length > 0) {
    foto = zitate_foto;
  }
}

/* ** Berechnen von `anderepaare` ** */
/* andere, bereits vorhandene frontmatter Einträge hinzufügen */
let attribs = tp.frontmatter;
let propertyNames = Object.keys(attribs);
let propertyName = "";
let propertyValue = "";
for (let ix = 0; ix < propertyNames.length; ix++) {
  propertyName = propertyNames[ix];
  propertyValue = tp.frontmatter[propertyName];
  if(propertyName != "aliases" && 
    propertyName  != "date_created" && 
    propertyName  != "tags" && 
    propertyName  != "publish" && 
    propertyName  != "cssclass" && 
    propertyName  != "position" ) {
      anderepaare += "\n" + propertyName + ": " + propertyValue
  }
}
_%>
aliases:      <% aliases %>
date_created: <% date_created %>
tags:         <% tags %>
publish:      <% publish %> 
cssclass:     <% cssclass %>
<%_ anderepaare %>
---

<%* if(foto.length > 0) { -%>
![<% dateityp %>](../_resources/<% foto %>)
<%* } _%>
<%* if(dateityp == "person") { -%>
* 0000-00-00 in Place, **†** 0000-00-00 in Place war/ist Mensch
<%* } _%>
<%* if(dateityp == "zitat") { -%>
Zitat

\- Autor, 1900
<%* } _%>

```

### For Download
[zwischen_Vorlage.md](https://github.com/SilentVoid13/Templater/files/11087177/zwischen_Vorlage.md)

**Comment by francisHugo** [comment]
hi @MonikaLobinger 

This seems like something what I am looking for to use in my obsidian vault.  I do not know any JS, neither the German languages, so it would be great if the explanations is in English.

meanwhile, what I am trying to do in my obsidian vault is:
1. when create a new note by press ctrl+N, if a note name starts with a prefix (such as [[prefix-bookName]], it will create a note based on a specific template and rename it to bookNote, and if possible, move it to a specific folder. These seem like what you called the create or modified note based on filetype. Without the prefix, it will create the note with the common template, and note is named to BookName;
2. if there is a [[prefix-bookName]] in a note, when click on the link, it will do the same, also, if there is [[bookName]] in a note.

I have been looking for such workflow for couple of months and find no results. 

Appreciate if you could help on this. Thanks in advance.

Francis

**Reply by MonikaLobinger** [reply]
Before answering I have to clarify:

- When creating a new note with Ctrl-N the note is new and empty and it has no name given by you. Obsidian names it  "Untitled" in English (and "Unbenannt" in German), subsequent new notes in the same folder would be named "Untitled1", "Untitled2" and so on.
- There are two ways of creating new notes with Templater
   - Create it with the command "Templater: Create new note from template" (Alt-N)
   - Give a folder template in the Templater settings
   - By the way: folder templates do not have to be stored in Templaters template folder. They can be stored in any folder. When assigning it to a folder, the full path and name including ".md" has to be specified, eg  to folder "root/folder" I assign the template "root/folder/folder_template.md"

### Folder Templates
My script is used as a folder template. I have a folder for books, persons, quotes, and something more (8 types), everything in one folder. Those types have different file markers, person and places do have none. In Templater settings I have assigned this script to this folder.

I could have 8 different templates in Templaters template directory. Then on creating a new Note, Templater would ask which template to use. 

But when using a folder template, I can only have one template. So the information which sort of note is created has to be retrieved from inside the script. tp.system.suggester(...) has a dropdown list with this 8 note types. 

When the script starts on new note creation in this folder it asks with this dropdown list, which note type should be created.
Then it asks with an other dialog what the note name should be. The file name (without marker) has to be entered.
Then it adds the marker, depending on the type to the notename and renames the note from "Untitled" or "Unbenannt" to e.g. "{b}author - title". Additionally it adds some frontmatter entries, but this seems not being your special interest.

### Changing existing files
With the command "Templater: Open insert template modal" (Alt-E) one can execute a template on an existing file. Templater asks in a dropdown list, which template to use. Only templates in the template folder are in the dropdown list. 

So, to use my script on existing files, I have to copy it in the template folder. Then I select it in the dropdown list.

Now my script knows the notename, it is an existing note. If it is a note with marker in the name, it already knows the note type. If not, it asks with a dropdown list for the type. It will not change the notename.


Executing templates with "Templater: Open insert template modal" (Alt-E) will create all output on cursor position. 

My script adds the new frontmatter header at cursor position. It is possible, to change the whole file content, but I am only starting with Templater programming. One has to delete the old frontmatter header afterwards manually.

### Moving files
It is possible to move files, I have read about this, but not yet worked with it.


### @francisHugo 
I know how difficult and disappointing it is, wanting to do something with Obsidian and not knowing how. Therefore I started to learn programming for Templater. But I programmed in other languages before. As you are not a programmer at all, this is much more difficult for you. I will help you as much as I can with my low Templater abilities. 

As in any programming task, one has to know  what is needed. A problem description has to be exact, so that a programm specification can be made, fullfilling the needs.

**Reply by francisHugo** [reply]
## Thank very much

Thanks very much for your detailed explanations. I am understanding somehow of your snippets, but still, I have to take some time to digest it.
In the other hand, I got something from the link #1055 and the example vaults I downloaded( https://github.com/Pamela-Wang/Obsidian-Starter-Vaults) several moths ago,  built an obsidian vault which almost meet my needs, except a problem in it. 

## within the vault, I can,

1. when create a new note by press Ctrl + N, use a prefix in the note name will have the note be created, renamed, also be able to be moved to certain folder ( with template snippets found in the Pamela-Wang's vault)
2. click on an *in-note uncreated named with prefix* link, will do the same as 1, except the link has no prefix.

## if in-note link has no prefix, there will be a problem

when such link is clicked, an input windows will be popped up to ask input the note name. so I have to re-input the note name again. I hope that it will directly use the link as the note name, not to re-input it again, moreover, if I put something else different from the exist link, it will rename the link and created the note with the new input name, and this is not what be expected and supposed to.

I enclosed the vault here, hope you could time to help me to solve this problem. Thanks in advance. Sorry for trouble you all of these.  

Also, currently, there are 2 templates use for one note type ( one template, one with "of meta" in its name), hope these 2 templates can be merged into one if possible.


[Test Template Vault.zip](https://github.com/SilentVoid13/Templater/files/11120911/Test.Template.Vault.zip)

**Comment by francisHugo** [comment]
Thanks for your detailed explanations. 
hope you may take a look at the vault I uploaded, and looking forward some helps from you, if possible. Thanks in advance.

**Reply by MonikaLobinger** [reply]
I have never created new notes by clicking on a  link to a non existing note.
Perhaps I now can understand your problems better.

**Reply by MonikaLobinger** [reply]
Hallo @francisHugo 

Exchange your "Meta Template Picker.md" with the appended version.
[Meta Template Picker.md](https://github.com/SilentVoid13/Templater/files/11128832/Meta.Template.Picker.md)

- You can remove the first line (output to console)
- If your Obsidian Language is neither German nor English you have to change "Unbenannt" in the 2nd line by the default filename for newly created files in your language.

Thank you for sharing example Vault, I got ideas out of it.

Best regards
Monika

**Reply by MonikaLobinger** [reply]
### Regarding merging two templates to one

> Also, currently, there are 2 templates use for one note type ( one template, one with "of meta" in its name), hope these 2 templates can be merged into one if possible.

It is possible to merge them in one. But it is complicated, as the chosen approach is not clean. If it were my code, I would rework it. I would not make aesthetic fixes for weird code - it gets from complicated to more complicated to no longer maintainable.

Nor will I do so in other peoples code. Sorry for that.

Edited: By the way
The code  I posted   [here in this thread](https://github.com/SilentVoid13/Templater/discussions/1071#discussion-5012458)  is weird code. I have fully reworked it. I will post it when I am sure, it's ok now.

**Reply by francisHugo** [reply]
Hi @MonikaLobinger

The **Meta Template Picker.md** works! it's so cool!, Thanks. Only except to switch the last two templates, or I would have no chances to input my prefix 😂 , anyway, it works now, thanks a lot!

As to merge the templates, I am used to it now. Since it is an only one-time works for ever, so it is OK for me. Never mind.

Looking forward your elegant codes to be posted soon.

Best Regards,
Francis

**Reply by MonikaLobinger** [reply]
Fine, that you like it
> Looking forward your elegant codes to be posted soon.
It will never get ready - now I am reworking it again

Best regards
Monika

**Reply by francisHugo** [reply]
Wow, that's terrific!

Hope it will be in English so I can read it. 😀

BR
Francis

**Comment by MonikaLobinger** [comment]
I have replace the script in my starting post with a 
[better version](https://github.com/MonikaLobinger/foty/blob/a488d4fcaa96accaed4156e7846246252f9121ed/_scripts/onne.js)  of it. It is used with [another template](https://github.com/MonikaLobinger/foty/blob/a488d4fcaa96accaed4156e7846246252f9121ed/_vorlagen/One_Vorlage.md)

This is the only template I use and it works for all files beside MOCs. 

But it is much to complicated and unstructured for my taste and it is  not generic - changing my needs or being used for other users needs will result in code changes (not only data changes)

So I am working on a generic version, learning much about javascript on the way. Coming to deep questions: "Does the key belong to value data or is it given from outside?" "Does a entry know its parent or is it only told?"

May be that I can not realize what I wish to have. May be this 3rd version will never become ready.

**Reply by francisHugo** [reply]
Congratulations! so fast!

one question, how does the scripts to be used with the **another template**?  it is too complicated to understand and difficult for me to try to use it.

after read it over and over, totally lost. 
 

Best Regards
Francis

**Reply by MonikaLobinger** [reply]
- copy the script in Templater script folder
- copy the [another template](https://github.com/MonikaLobinger/foty/blob/a488d4fcaa96accaed4156e7846246252f9121ed/_vorlagen/One_Vorlage.md) in template folder
- assign the template as folder template for Vault root folder in Templater settings
- remove all other Templater folder settings

But the script is only for my needs. Those are:
- only create new files (not by clicking on links)
- create new files in "current folder"  (there is an obsidian setting, where new notes should be placed)
- I have a mixture between marker settings and folder settings
- my 7 markers are: ,{a}, {b}, {v}, {w}, °, °°,$, - exactly the characters between the commas
- The folders which use markers are: "exzerpte" and "zwischenreich"
   - "exzerpte" uses one marker: `$`
   - "zwischenreich" uses the other 6
   - So if you create a note in "exzerpte" it will start with `$` and the name you enter in the dialog "which name", which questions for name (don't give the marker - the opposite of your approach)
   - If you create a note in "zwischenreich" you will be asked two questions: "Which type" and "which name", where "which type" is a selection dialog containing the 6 types
- folders which create diary entries are "diary", "gartentagebuch", "lesetagebuch" and "unbedacht"
- those names are the deepest folder in a folder structure, they have not to be root folders (So "exzerpte" may be placed in "VAULTROOT/trythisscript/exzerpte"

You see, my needs are others then yours, no copiing around. 

And this is one of the reasons why I try a 3rd approach 

PS: This 2nd version is ready for 5 days or so, but it is not really good

**Reply by francisHugo** [reply]
Wow! more details info coming! thanks.

I also visited your repository, the foty, seems you are building wonders! 

I hope that I would be the first one of witness the wonders happen, and also be happy to make the wonders happen in my Obsidian vault, too.

However, with my little knowing of JS, and zero programming skills, it sure will take much more times for me to learn and to practice it.

anyway, the foty I could understand a little with your clearly explanations, I will learn how to use it soon.

questions are coming soon 😀
Besrt Regards,

Francis

**Reply by MonikaLobinger** [reply]
It is not yet ready - work in progress
